### ISSUE TITLE
Ringed code handles default values differently than unringed code

### ISSUE THREAD

**Issue by gigamonkey** [jmoenig/Snap#3006]
Define this simple block:

![image](https://user-images.githubusercontent.com/250053/157910927-816c9a57-a200-43e8-88b8-a42ba2cb3d6f.png)

And this one:

![image](https://user-images.githubusercontent.com/250053/157911083-64901b8d-43b1-4614-98ac-8b369516761b.png)

Then observe:

![image](https://user-images.githubusercontent.com/250053/157911327-5762bf8a-5a7a-4a8b-a535-ec088dba4a9d.png)

vs.

![image](https://user-images.githubusercontent.com/250053/157911419-8badc421-a94c-4333-9182-55ff68c2ea66.png)

With explicit args both the direct invocation and the `CALL RING` version report the same result:

![image](https://user-images.githubusercontent.com/250053/157911244-3b342a2b-ccdb-4f40-bd7f-793872eefcb1.png)

![image](https://user-images.githubusercontent.com/250053/157911521-5b53f044-964f-4553-a975-feb9b7b89356.png)

and

![image](https://user-images.githubusercontent.com/250053/157911275-f9ce8c72-5aa9-4758-ae14-055dd837c835.png)

![image](https://user-images.githubusercontent.com/250053/157911572-4a0b85cb-2449-4c52-83b3-532c418f0e83.png)

I have a suspicion that this is perhaps related to the issue underlying this thing from the forums:

https://forum.snap.berkeley.edu/t/split-of-a-block-not-equal-to-the-list/10189

**Comment by cycomachead** [comment]
I haven’t tested this yet, but isn’t this a function of the numeric input type?

-- 
Michael Ball
From my iPhone
michaelball.co

**Comment by gigamonkey** [comment]
I assume so. But my expectation was that the `TEST` block would behave the same regardless of whether it was run directly or via `CALL`.

**Comment by brianharvey** [comment]
TEST reports 0 because, _all else being equal,_ an empty numeric slot defaults to 0.  If the input were of type Any, it'd report an empty string.

But in the context of CALL (or any other higher order function), an empty slot has a different meaning: it's an implicit formal parameter, into which an actual argument of the call will be substituted:
![untitled script pic](https://user-images.githubusercontent.com/3885074/157992946-0b868547-47a5-4158-a9fc-5da1d80b6d0c.png)
So if you rewrite CALL RING this way:
![untitled script pic (1)](https://user-images.githubusercontent.com/3885074/157993373-0ce98993-baed-4fcc-85a9-2781f924809f.png)
you can see that feature at work:
![untitled script pic (2)](https://user-images.githubusercontent.com/3885074/157993422-2114a33f-37eb-4f2c-b0b7-02a8ed84162a.png)
And if you _don't_ supply an actual argument, the empty slot is filled with nothing.

Yes, this is a problematic conflicting meaning in different contexts.  The default-to-zero behavior is something we inherited from Scratch; if I were designing Snap<i>!</i> from scratch instead of from Scratch we wouldn't have that default, and an empty numeric slot would always mean that you want something substituted into it.  But I imagine Jens would disagree.

The use of an empty slot for input substitution is something I'm really proud of, because it means a beginner kid can readily understand things like
![untitled script pic (3)](https://user-images.githubusercontent.com/3885074/157993856-6e1c05b9-0601-4321-b3d1-22a3d640e4ed.png)
![untitled script pic (4)](https://user-images.githubusercontent.com/3885074/157993935-aed0498c-89b3-4184-ac69-47ba155a654d.png)
without _really_ knowing anything about lambda.

The root of this feature is the following piece of developmental psychology:  If you ask an eight-year-old "x+3=7; what's x?" the kid will probably just stare at you blankly, but if you ask the same kid "⃞+3=7; what number goes in the box?" they'll say "four."  So the empty input slot is our version of ⃞.

I've tried convincing Jens that we should decorate the empty slots when they're in a context in which something will be substituted:
![circles3](https://user-images.githubusercontent.com/3885074/157995269-c2777312-e124-4f47-8ee2-f864f4d673e7.png)
but he just yells at me.  :/

**Comment by cycomachead** [comment]
I think the question is why is `call` with no arguments different than just calling the block directly. 

![untitled script pic-2](https://user-images.githubusercontent.com/1505907/157997400-dd45e2fe-f9c7-4c60-a4a8-52337c10c8e0.png)
![untitled script pic-3](https://user-images.githubusercontent.com/1505907/157997440-93f1871e-e953-47c9-a523-4b949e99c557.png)

I can reason why the 2nd image happens - in that we bypass the numeric coercion, but the first example is a little weird. I guess perhaps the real issue is that we are calling a 1 argument function with zero arguments...

**Comment by brianharvey** [comment]
Yeah, the case of one empty slot and zero actual arguments may be a bug.

**Comment by brianharvey** [comment]
Well, I guess we could make sure the specific case of zero actual inputs doesn't try to substitute into empty slots...

**Comment by cycomachead** [comment]
Ah that’s very interesting. 

I’m not sure changing the behavior of no inputs is really safe - it seems like it could be relied upon in many places. And that in Snap! we don’t enforce passing data to a function - because you can’t distinguish between no argument provided and the empty string (or even nil). But the coercion maybe is something that we could address. Hmm.

**Comment by brianharvey** [comment]
The opposite situation -- providing inputs to a niladic block -- is handled specially, a fairly recent change, because otherwise there's no good way to provide an _input list_ in the call to a variadic block.  When you're writing the code you don't know how big that input list will be, so you can't provide a ringed block with exactly the right number of input slots.  So we made a niladic ringed block accept nonempty input lists.

But I can't offhand think of a case in which this case's behavior would be useful.

**Comment by jmoenig** [comment]
This has nothing to do with "the context of CALL". It's all about rings. A ring turns an expression into a function. You can expand the ring to specify formal parameters. If you *don't* expand it all *empty input slots* of the expression turn into implicit formal parameters of the function. This happens when the function is *made*, not when it's called. You can assign the result of evaluating a ring to a variable / parameter and then CALL it somewhere else and some other time. What's quirky about implicit parameters is that Snap doesn't decide until CALL how to distribute the arguments you pass into it. If you give it a single argument, it will be filled into *every* empty slot, whereas if you give it *more than one* argument, they will be assigned to the empty inputs slots in their lexical order, and Snap requires their numbers to match. This is both really cool and highly problematic, because we cannot properly reason about functions in Snap. We can't even tell - just by looking at a function - how you're supposed to call it, i.e. how many (and which) parameters it has. Unless you use explicit ones by expanding the ring.

**Comment by brianharvey** [comment]
>This has nothing to do with "the context of CALL". It's all about rings. A ring turns an expression into a function.

Yes, okay, much of the work happens in the ring.  But it doesn't _matter_ what slots Snap<i>!</i> considers substitutable until some substitution actually happens, namely, in CALL (or other HOFs).  If you ringify an expression but never actually call it, then there's no way for the user to know, and no reason for the user to care, that you've done some analysis of empty slots when the expression was ringified.

> We can't even tell - just by looking at a function - how you're supposed to call it, i.e. how many (and which) parameters it has.

Exactly.  Which is why, I still say, it's in the CALL (or other primitive HOF) that empty slots become argument slots.  Not until the CALL do we know how many actual arguments the user will supply.  It can't (as you point out) even be done by pre-runtime syntactic analysis of the CALL and its ringed function, if the user has used the INPUT LIST: option, and you don't know the length of the list.

So, yeah, you do some optimization by pre-parsing the expression in the ring and remembering where its empty slots are.  But that's a secondary issue.

**Comment by brianharvey** [comment]
But, getting back on topic, is there any reason we want a call to a monadic function of a numeric input slot, called with no inputs, to prevent the slot having its default value of zero?

**Comment by jmoenig** [comment]
Snap has several values that when applied to numeric operations qualify as zero, among others, the empty string, a string containing only blank spaces or any of JavaScripts zero-ish/falsey values, mostly for reasons of performance. Since we don't really have strong types that's okay for Snap to get away with

![zeroish1](https://user-images.githubusercontent.com/2596285/158013442-259c5546-2281-40af-8291-73cd5818c4b0.png) 
![zeroish2](https://user-images.githubusercontent.com/2596285/158013444-a4441f7b-a758-4c83-978a-7a56cb5109d9.png)

until folks think of themselves al mathematicians and muse about the metaphysics of equality :)

**Comment by jmoenig** [comment]
If you ringify an expression but never actually call it, then there's no way for the user to know, and no reason for the user to care, that you've done some analysis of empty slots when the expression was ringified.

> If you ringify an expression but never actually call it, then there's no way for the user to know, and no reason for the user to care, that you've done some analysis of empty slots when the expression was ringified.

I disagree. Not being able to reason about the syntax of a function is a serious deficiency of the language and probably the single most serious design flaw of Snap.

**Comment by brianharvey** [comment]
I'm sorry, but I can't figure out how what you said is a response to what I said.  This is probably me being stupid, but could you please explain the connection?

**Comment by gigamonkey** [comment]
Just to check my understanding when I write this block:

![image](https://user-images.githubusercontent.com/250053/158032113-b42f7277-8e00-4a80-bb81-0b53b658b159.png)

I assume that is basically equivalent to (pardon my Common Lisp):

```
(defun test (&optional (n 0)) n)
```

So then the question is what does:

![image](https://user-images.githubusercontent.com/250053/158032295-58f4356c-b54f-424d-a15c-f490310989ea.png)

translate to.

I was, in the context I actually ran into this problem, expecting/wanting it to be: 

```
(lambda () (test))
```

I.e. just invoke my block without passing it any arguments. But that's no good, as @brianharvey explained, because the ring needs to produce a function that can be passed arguments and then needs to use the arguments to fill in the hole(s) in the ringed expression, so it is more like:

```
(lambda (&optional x) (test x))
```

The problem is this ring has a generic optional parameter and therefore `x` defaults to `nil` (in Common Lisp) which is then explicitly passed to `test` so the default value for `n` in test is not used and the value that is used isn't of the right type for the hole.

In Common Lisp I could implement the ring more transparently like this:

```
(lambda (&optional (x nil x-supplied-p))
  (if x-supplied-p (test x) (test)))
```

(If you're not familiar with all the nooks and crannies of Common Lisp `&optional (x nil x-supplied-p)` is actually declaring one optional parameter named `x` which defaults to the given value, `nil`, and also causing another variable `x-supplied-p` to be created that will be true when a value was explicitly provided for `x` and false otherwise.)

In which case 

![image](https://user-images.githubusercontent.com/250053/158032456-0d4ba44e-0ce1-4789-9200-dfee8a458836.png)

would result in the `CALL` not passing any arguments to the ring which would then invoke `TEST` with no arguments and let its own default argument value come into play.

I suppose (and maybe this is what you're saying @jmoenig?) that in an ideal world the ring would be able to walk the thing it is ringing and figure out how many holes there are and maybe what types they are expecting and then produce an appropriate custom lambda?

I.e. if I wrote a block that is the equivalent of:

```
(defun some-block (&optional (num 0) (list ()))
   ;; i.e. a block with one numeric and one list argument
)
```

then a ring placed around an instance of that black could inspect it's parameters and translate into something like:

```
(lambda (&optional (num 0) (list ())) (some-block num list))
```

Is that at all right?

**Comment by brianharvey** [comment]
So you're saying we should strive to be a Common Lisp instead of a Scheme?  ;-)

(Actually in one sense we already are:  We're a Lisp-2, not a Lisp-1.)

I think what you said is right, except that we all, even Jens, tell people that types are advisory in Snap<i>!</i>, and that's mostly true, except for this default-to-zero behavior of Number slots.  In my idea of a perfect world, all input slots would default to #undefined or something, and the analysis of code would be much easier.

**Comment by gigamonkey** [comment]
> So you're saying we should strive to be a Common Lisp instead of a Scheme? ;-)

I'm not _not_ saying that. ;-)

**Comment by brianharvey** [comment]
The idea is that if you provide explicit formal parameters in a ring, you are taking charge of argument placement, so Snap<i>!</i> doesn't substitute into any empty slots.
